presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -89,251 +180,569 @@ private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocato
+            }
+
+            // add new projections for hash symbols needed by the parent
+            for (List<Symbol> partitionSymbols : context.getPreferredHashSymbols()) {
+                if (!result.getHashSymbols().containsKey(partitionSymbols)) {
+                    Expression hashExpression = getHashExpression(partitionSymbols);
+                    Symbol hashSymbol = symbolAllocator.newHashSymbol();
+                    assignments.put(hashSymbol, hashExpression);
+                    outputHashSymbols.put(partitionSymbols, hashSymbol);
+                }
+            }
+
+            return new PlanWithProperties(new ProjectNode(idAllocator.getNextId(), source, assignments.build()), outputHashSymbols);
+        }
+
+        private PlanWithProperties plan(PlanNode node, Context context)
+        {
+            PlanWithProperties result = node.accept(this, context);
+            checkArgument(
+                    result.getNode().getOutputSymbols().containsAll(result.getHashSymbols().values()),
+                    "Node %s declares hash symbols not in the output",
+                    result.getNode().getClass().getSimpleName());
+            return result;
         }
     }
 
-    private static ProjectNode getHashProjectNode(PlanNodeIdAllocator idAllocator, PlanNode source, Symbol hashSymbol, List<Symbol> partitioningSymbols)
+    private static class PlanWithProperties
     {
-        checkArgument(!partitioningSymbols.isEmpty(), "partitioningSymbols is empty");
-        ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
-        for (Symbol symbol : source.getOutputSymbols()) {
-            Expression expression = new QualifiedNameReference(symbol.toQualifiedName());
-            outputSymbols.put(symbol, expression);
+        private final PlanNode node;
+        private final BiMap<List<Symbol>, Symbol> hashSymbols;
+
+        public PlanWithProperties(PlanNode node, Map<List<Symbol>, Symbol> hashSymbols)
+        {
+            this.node = requireNonNull(node, "node is null");
+            this.hashSymbols = ImmutableBiMap.copyOf(requireNonNull(hashSymbols, "hashSymbols is null"));
+        }
+
+        public PlanNode getNode()
+        {
+            return node;
+        }
+
+        public BiMap<List<Symbol>, Symbol> getHashSymbols()
+        {
+            return hashSymbols;
         }
 
-        Expression hashExpression = getHashExpression(partitioningSymbols);
-        outputSymbols.put(hashSymbol, hashExpression);
-        return new ProjectNode(idAllocator.getNextId(), source, outputSymbols.build());
+        public Symbol getRequiredHashSymbol(List<Symbol> symbols)
+        {
+            Symbol hashSymbol = hashSymbols.get(symbols);
+            requireNonNull(hashSymbol, () -> "No hash symbol generated for " + symbols);
+            return hashSymbol;
+        }
+    }
+
+    private static Map<Symbol, Symbol> computeIdentityTranslations(Map<Symbol, Expression> assignments)
```

Comment:
This code seems to be replicated in a bunch of places (`AddExchanges`, `PropertyDerivations`). Can we reuse this?